File: app/schemas/ui.py
from __future__ import annotations

from pydantic import BaseModel, Field


class SessionListItem(BaseModel):
    session_id: str
    user_id: str | None = None
    title: str | None = None
    message_count: int
    last_message_at: str | None = None
    created_at: str | None = None


class SessionsResponse(BaseModel):
    sessions: list[SessionListItem]


class SessionUpdateRequest(BaseModel):
    title: str | None = Field(default=None, max_length=200)
    is_active: int | None = Field(default=None, ge=0, le=1)


class ConversationMessage(BaseModel):
    id: str
    role: str
    content: str
    sequence_number: int
    created_at: str | None = None
    tool_name: str | None = None
    token_count: int | None = None


class SessionMessagesResponse(BaseModel):
    session_id: str
    messages: list[ConversationMessage]


class ToolActivityItem(BaseModel):
    session_id: str
    message_id: str
    tool_name: str
    created_at: str | None = None
    role: str


class ToolActivityResponse(BaseModel):
    activities: list[ToolActivityItem]


class MemorySummaryResponse(BaseModel):
    session_id: str
    recent_user_messages: int
    recent_assistant_messages: int
    total_messages: int


class SystemStatusResponse(BaseModel):
    app_name: str
    app_env: str
    uptime_seconds: int
    model_name: str | None = None
    load_avg_1m: float | None = None
    load_avg_5m: float | None = None
    load_avg_15m: float | None = None
    sqlite_path: str
    chroma_path: str


class UsageSummaryResponse(BaseModel):
    user_id: str
    total_messages: int
    assistant_messages: int
    total_tokens: int


class UserProfile(BaseModel):
    name: str | None = None
    language: str | None = None
    location: str | None = None
    profession: str | None = None
    project: str | None = None
    preferences: list[str] = []


class UserProfileResponse(BaseModel):
    user_id: str
    profile: UserProfile


class ThinkingStep(BaseModel):
    key: str
    label: str
    status: str
    detail: str | None = None


class ThinkingProcessResponse(BaseModel):
    session_id: str
    steps: list[ThinkingStep]


class SystemMetricsResponse(BaseModel):
    cpu_percent: float
    memory_percent: float
    gpu_percent: float | None = None
    npu_percent: float | None = None


class SystemOverviewResponse(BaseModel):
    """Single round-trip for dashboard: metrics + status."""

    metrics: SystemMetricsResponse
    status: SystemStatusResponse


class FileEntry(BaseModel):
    name: str
    path: str
    is_dir: bool
    size: int | None = None
    modified_at: str | None = None


class FileListResponse(BaseModel):
    root: str
    current_path: str
    entries: list[FileEntry]


class FileContentResponse(BaseModel):
    path: str
    content: str


class FileSearchMatch(BaseModel):
    path: str
    line_number: int
    line: str


class FileSearchResponse(BaseModel):
    query: str
    matches: list[FileSearchMatch]


class MessageFeedbackRequest(BaseModel):
    value: str = Field(..., pattern="^(like|dislike|none)$")


class MessageFeedbackResponse(BaseModel):
    message_id: str
    value: str


class VoiceStatusResponse(BaseModel):
    state: str
    audio_level: float
    detail: str | None = None
    updated_at: float


class MeResponse(BaseModel):
    user_id: str
    display_name: str | None = None
    language: str | None = None
    location: str | None = None
    profession: str | None = None


class CapabilitiesResponse(BaseModel):
    chat: bool
    voice_chat: bool
    memory_search: bool
    sessions: bool
    session_restore: bool
    session_permanent_delete: bool
    thinking_process: bool
    thinking_stream: bool
    files: bool
    tools_activity: bool
    system_status: bool
    system_metrics: bool
    system_overview: bool
    profile: bool
    me: bool
    message_feedback: bool

File: frontend/src/types/ui.ts
export type Session = {
  session_id: string;
  title: string | null;
  user_id?: string | null;
  message_count?: number;
  last_message_at?: string | null;
  created_at?: string | null;
};

export type Message = {
  id: string;
  role: string;
  content: string;
  created_at?: string | null;
  token_count?: number | null;
};

export type ChatResponse = {
  session_id: string;
  user_message_id: string;
  assistant_message_id: string;
  response: string;
  used_memory: boolean;
  used_internet: boolean;
  audio_path?: string | null;
};

/** `POST /voice-chat` JSON — play TTS via `GET {apiBase}/{audio_url}`. */
export type VoiceChatResponse = {
  session_id: string;
  transcript: string;
  response: string;
  used_memory: boolean;
  used_internet: boolean;
  audio_path?: string | null;
  audio_url?: string | null;
  /** Voice ignored (wake gate); no new chat messages. */
  skipped?: boolean;
  /** wake_gate | no_speech */
  skip_reason?: string | null;
  voice_listen_paused?: boolean | null;
  voice_wake_session_active?: boolean | null;
};

export type Metrics = {
  cpu_percent: number;
  memory_percent: number;
  gpu_percent: number | null;
  npu_percent: number | null;
};

export type SystemStatus = {
  app_name: string;
  app_env: string;
  uptime_seconds: number;
  model_name?: string | null;
  load_avg_1m: number | null;
  load_avg_5m: number | null;
  load_avg_15m: number | null;
  sqlite_path: string;
  chroma_path: string;
};

export type ThinkingStep = {
  key: string;
  label: string;
  status: string;
  detail?: string | null;
};

export type ThinkingProcess = {
  session_id: string;
  steps: ThinkingStep[];
};

export type ToolActivity = {
  session_id: string;
  message_id: string;
  tool_name: string;
  created_at: string | null;
  role: string;
};

export type ToolActivityListResponse = {
  activities: ToolActivity[];
};

export type Profile = {
  name: string | null;
  language: string | null;
  location: string | null;
  profession: string | null;
  project: string | null;
  preferences: string[];
  /** Per-user TTS tempo (e.g. 1.2). Omitted = server default. */
  tts_playback_speed?: number | null;
  /** Say this name in voice when silent mode is on. */
  assistant_wake_name?: string | null;
  /** When true, voice is ignored unless the transcript includes the wake name (text chat always works). */
  voice_listen_paused?: boolean;
  /** After wake in quiet mode: follow-up voice works without wake until stop / keep quiet. */
  voice_wake_session_active?: boolean;
};

export type ProfileResponse = {
  user_id: string;
  profile: Profile;
};

export type UsageSummary = {
  user_id: string;
  total_messages: number;
  assistant_messages: number;
  total_tokens: number;
};

export type VoiceStatus = {
  state: string;
  audio_level: number;
  detail: string | null;
  updated_at: number;
};

export type MessageFeedbackValue = "like" | "dislike" | "none";

export type MessageFeedbackResponse = {
  message_id: string;
  value: MessageFeedbackValue;
};

export type FileListResponse = {
  root: string;
  current_path: string;
  entries: {
    name: string;
    path: string;
    is_dir: boolean;
    size?: number | null;
    modified_at?: string | null;
  }[];
};

export type FileContentResponse = {
  path: string;
  content: string;
};

export type MemorySearchMatch = {
  message_id: string;
  session_id: string;
  role: string;
  content: string;
  score: number;
  created_at?: string | null;
};

export type MemorySearchResponse = {
  matches: MemorySearchMatch[];
};

export type FileSearchMatch = {
  path: string;
  line_number: number;
  line: string;
};

export type FileSearchResponse = {
  query: string;
  matches: FileSearchMatch[];
};
